gitcache/main.go

```diff
@@ -4,139 +4,191 @@ import (
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
 	}
 
-	// opening the blob
-	bucket, err := New(blob)
+	sugar.Info("Storing the last commit ", commit.Author.When)
+	// storing the last commit to the blob
+	err = bucket.Set(storage.BlobLastCommitPath, []byte(fmt.Sprintf("%64b", commit.Author.When.Unix())))
 	if err != nil {
 		log.Panic(err)
 	}
 
-	archivePath := path.Join(tarDir, fmt.Sprintf("%s.tar.gz", repo.Repo))
-	// storing the last commit to the blob
-	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastcommit", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", commit.Author.When.Unix())))
+	data, err = archiveFolder(storage.GitDir, storage.GitTarFile)
 	if err != nil {
 		log.Panic(err)
 	}
 
-	now := time.Now().Unix()
-	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/lastsync", repo.Owner, repo.Repo), []byte(fmt.Sprintf("%64b\n", now)))
+	err = bucket.Set(storage.BlobGitFolderPath, data)
 	if err != nil {
 		log.Panic(err)
 	}
 
 	// removing the .git folder as it is not required for the tar ball
-	err = os.RemoveAll(path.Join(dir, ".git"))
+	err = os.RemoveAll(path.Join(storage.GitDir, ".git"))
 	if err != nil {
 		log.Panic(err)
 	}
 
-	// creating an archive
-	archive := archiver.NewTarGz()
-
-	err = archive.Archive([]string{dir}, archivePath)
+	data, err = archiveFolder(storage.GitDir, storage.BlobArchiveFile)
 	if err != nil {
 		log.Panic(err)
 	}
 
-	data, err := ioutil.ReadFile(archivePath)
+	err = bucket.Set(storage.BlobArchivePath, data)
 	if err != nil {
 		log.Panic(err)
 	}
-
-	// storing the archive to the blob
-	err = bucket.Set(fmt.Sprintf("gitcache/%s/%s/tar", repo.Owner, repo.Repo), data)
+	err = bucket.Set(storage.BlobLastSyncPath, []byte(fmt.Sprintf("%64b", time.Now().Unix())))
 	if err != nil {
 		log.Panic(err)
 	}
+	sugar.Info("Finished processing for the first time ", repo)
 }
 
-type RepoURL struct {
-	Host, Owner, Repo string
+// clones the repository if it does not exists. If the git repo already exists then it fetches from the remote.
+func cloneGitRepo(storagePath *StoragePath, repo RepoURL) (*git.Repository, error) {
+	// Clone the given repository to the given directory
+	gitRepo, err := git.PlainClone(storagePath.GitDir, false, &git.CloneOptions{
+		URL: fmt.Sprintf("http://%s/%s/%s", repo.Host, repo.Owner, repo.Repo),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to clone %+v", repo)
+	}
+	return gitRepo, nil
 }
 
-func (r *RepoURL) String() string {
-	return fmt.Sprintf("%s/%s/%s", r.Host, r.Owner, r.Repo)
-}
+func archiveFolder(folderToArchive, archivePath string) ([]byte, error) {
+	tarformat := archiver.DefaultTarGz
+	err := tarformat.Archive([]string{folderToArchive}, archivePath)
+	if err != nil {
+		return nil,
+			errors.Wrapf(err, "unable to read the archive the folder %s to the archive path %s", folderToArchive, archivePath)
+	}
 
-func (r *RepoURL) Type() string {
-	return "repo"
+	data, err := ioutil.ReadFile(archivePath)
```

Comment:
This doesn't seem like it scales if we are dealing with very large repos. What if we try to clone e.g. a browser like chromium (which is several GBs)? 

Can we just write them to disk somewhere and upload to GCS that way?

Reply:
Agree, it might not be able to scale very well.  What is your suggestion with the go-cloud package?

Reply:
Looks like the gcsblob package that you are using does not support this, so I would recommend using the main gcs package, which provides methods using: https://pkg.go.dev/cloud.google.com/go/storage#ObjectHandle.NewWriter and io.Copy or something similar.

Reply:
I did my research and couldn't find it in the gocloud package. 

We are trying to keep this open across cloud vendors and that is the reason for choosing that package.  

So this would make it only for GCP which is something we should avoid.

Reply:
Looks that is specific to GCP. We would like to keep it open across cloud vendors.

Reply:
That's a good point. Looking more closely at go-cloud, looks like they offer this equivalent feature instead: https://pkg.go.dev/github.com/google/go-cloud@v0.7.0/blob#Bucket.NewWriter. Can we use this?

Reply:
Cool, Yes I think it is possible. Let me give it a try. Thanks

Reply:
The underlying library that the code is using for `tar` and `untar` doesn't support IO writer. The standard `tar` and `untar` funcs have a problem in handling symlinks. 

I agree this is important. But let's take this as an action item and address it.

Reply:
OK, let's address this in a future PR. I guess we should be able to use that library to write to a file on disk first, and then uploading that way right?

Reply:
Yes, It should.